RavenDB-16614 Atomic read write in cluster wide transaction without c…

ayende opened the pull request:
…onflicts.

Not done yet - allowing the user to change from the default.
Properly testing deletes.

ayende committed:
RavenDB-16614 Atomic read write in cluster wide transaction without conflicts.

karmeli87 commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
+                                Index = putIndex,
+                                Document = context.ReadObject(new DynamicJsonValue {["Id"] = docId}, "cmp-xchg-content")
+                            });
+                            if (metadata != null) // just to be on the safe side, should never happen
```
why this should never happen ?

karmeli87 replied:
What would happened if the `metadata` is `null`, should we create a new `metadata`?

ayende replied:
If this happens, that means that the user submitted a document without metadata.
The only way that can happen is if they crafted a special request. I'm including this check here to avoid NRE in that case, but this isn't something likely.

karmeli87 commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -66,6 +66,7 @@ public async Task BulkDocs()
                     BatchTrafficWatch(command.ParsedCommands);
                 }
 
+                var disableAtomicDocumentWrites = GetBoolValueQueryString("disableAtomicDocumentWrites", required: false) ?? false;
```
How does the user set the `disableAtomicDocumentWrites` option?
to maintain compatibility, maybe we need to allow setting this globally on the `DocumentStore`?

ayende replied:
Currently, they don't, haven't gotten to it

ayende replied:
And yes, need to enable that per session & globally.

karmeli87 commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
+                                // this can happen if the user tried to delete without first loading the document
+                                // we can safely assume that such a blind write doesn't care about the state of the document
+                                // and just proceed normally
+                                deleteIndex = -1;
```
This is actually a problem.. I think.

If I delete and modify concurrently I might get successful response for both, while the document might or might not exist and we back to the original issue.

I think it would be better to enforce loading the document first (check on the client side as well?).

ayende replied:
If you call `Delete(string id)`, that is a good indication that you don't care what the current value is.

ayende committed:
RavenDB-16614 Fixing behavior when deleting documents in cluster wide transaction

ayende committed:
RavenDB-16614 Allow to disable the default of atomic document writes using server / database wide configuration, per document store or for a particular session

ppekrol commented:
Rebased

ayende committed:
RavenDB-16614 Fix commands should not have settable fields / properties.

karmeli87 reviewed (commented):
LGTM, but what will happened if...

1. mixing cluster-tx with normal I think it shouldn't be an issue.. IIRC we always prefer the document of the cluster-tx
2. conflicts or updates because of an external replication... we are replicating the document but not the compare exchange values.. which might be a problem?

ayende commented:
The document contains a `@cluster-tx-id` field, which is set to the index of the compare exchange value.
In such a case, we'll have a problem, yes. I'm not sure what would be a good way to resolve it.

If the field is now out of sync of the compare exchange, what should we do?

For that matter, we probably need to strip this value on external replication, because that is a per cluster setting.

We can try to do a repair on the cluster transaction value, I guess? But I'm not sure how viable that would be.

karmeli87 commented:
Maybe we can add the TopologyId to the compare-exchange name and to the metadata to make it some sort of multi-cluster change vector.
`@multi-cluster-change-vector`:`<topology1>:<raft-index>, <topology2>:<raft-index>`

One issue that I can think with this approach is in import to new database, I think.

ayende commented:
That wouldn't help the scenario of misaligned updates, thoughs.

ayende commented:
Consider this scenario:

* I load a document with `cluster-tx-id` of 7 in tx A.
* Modify that in a cluster tx B.
* The compare exchange value is now 8.
* I call `SaveChanges()` on tx A, overwriting the old value and returning the *`cluster-tx-id` to be. 
* All cluster wide operations will fail now on that document.

ryanheath commented:
Nice to see this is coming to RavenDB, making cluster wide transactions a whole lot easier to work with!
Mad props @ayende !

ryanheath commented on test/SlowTests/Issues/RavenDB-16614.cs:
```diff
@@ -0,0 +1,179 @@
+                var user2 = await session.LoadAsync<User>("users/phoebe");
+                user2.Name = "Phoebe Eini";
+
+                using (var conflictedSession = store.OpenAsyncSession(new SessionOptions {TransactionMode = TransactionMode.ClusterWide}))
```
How can we make sure that this session is hitting another node? Or is this done implicit by the test runner?
If implicit, how can we make sure the test also succeeds with hitting one node each time?

ayende replied:
We aren't trying to do that. Cluster Wide transactions are going to go to the leader anyway, so it doesn't matter from our perspective.

ppekrol commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -190,15 +200,84 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
             var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
             var commandsCountPerDatabase = context.Transaction.InnerTransaction.ReadTree(ClusterStateMachine.TransactionCommandsCountPerDatabase);
-            var commands = SerializedDatabaseCommands.Clone(context);
+            var commands = context.ReadObject(SerializedDatabaseCommands, "serialized-tx-commands");
```
Is this change needed?

ayende replied:
Yes, we are mutating the `SerializedDatabaseCommands` to add artifical commands here.

ayende committed:
Merge branch 'v5.2' into RavenDB-16614

ayende committed:
RavenDB-16614 - Fixing NRE on commit without documents
- Adding better error if we leak TcpConnectionOptions

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -58,6 +58,8 @@ public abstract partial class InMemoryDocumentSessionOperations : IDisposable
 
         private BatchOptions _saveChangesOptions;
 
+        public bool DisableAtomicDocumentWritesInClusterWideTransaction;
```
internal readonly?

ayende replied:
No, this is so the user can disable that for a single session.

ppekrol commented on src/Raven.Server/Documents/TcpHandlers/TcpConnectionOptions.cs:
```diff
@@ -114,9 +114,10 @@ public void Dispose()
         }
 
 #if !RELEASE
+        private string ss = Environment.StackTrace;
```
Can we remove this?

ayende replied:
Yes, I added this to see if it would fail with additional information.

ppekrol commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -69,6 +70,7 @@ public class ClusterTransactionOptions : IDynamicJson
             public TimeSpan? WaitForIndexesTimeout;
             public bool WaitForIndexThrow;
             public List<string> SpecifiedIndexesQueryString;
+            public bool DisableAtomicDocumentWrites;
```
This will be false for legacy commands? Is this OK?

ayende replied:
I don't think that this is a scenario that we need to worry about.

ppekrol reviewed (commented):
@ayende can you check failing tests?

ayende committed:
RavenDB-11795 - Fixing failing test (culture instance aren't the same)

ayende committed:
RavenDB-16614 - Fixing tests

ayende committed:
RavenDB-16614 Better error message when we fail a transaction

ayende committed:
RavenDB-16614
- Making sure that we are generating db group specific values for each transaction
- Prevent saving documents with the wrong cluster transaction index for our topology, which should alleviate issues with mismatched versions
- Better error when we are unable to save a document due to atomic guard mismatch.
- Fixing metadata save of complex objects

ayende commented:
Here is what the document will look like:

```
{
    "Name": "arava",
    "@metadata": {
        "@collection": "Users",
        "Raven-Clr-Type": "SlowTests.Issues.RavenDB_16614+User, SlowTests",
        "@flags": "FromClusterTransaction",
        "@cluster-transaction-index": {
            "Au3CeLCT8kWAU1UB4QcDuQ": 11
        }
    }
}
```

ayende committed:
RavenDB-16614 - Avoid @ prefix for the cluster index metadata, this may be stripped by older clients.

garayx commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -15,10 +15,21 @@ namespace Raven.Client.Documents.Commands.Batches
+            base.AppendOptions(sb);
+            if (DisableAtomicDocumentWrites)
+            {
+                sb.Append("?disableAtomicDocumentWrites=true");
```
should be `&disableAtomicDocumentWrites=true` if ` _options` are not null?

ayende replied:
I don't follow?

ppekrol replied:
At the end if null then we should not send anything (and then use what s configured in the database), if not null then send whatever is configured because client will override it.

ayende committed:
RavenDB-16614 - Making sure that we can default to no operation, so server admin can define a default and client can decide to override it locally

ayende committed:
RavenDB-16614
- Handling the scenario where we failed to @refresh a document
- Do not fail on cluster transaction index mismatch when accepting the data over replication, since it proceed at a different pace from the cluster.

ayende committed:
RavenDB-16614
- When saving a document with atomic guard metadata (either new version or the existing one) - validate that the atomic guard index matches.
- When importing / restoring from backup - update the atomic guard documents to match the new index on the new topology. If there are missing documents, remove the orphaned guards.

ayende committed:
RavenDB_16614 - Fixing tests

ayende commented:
Adding the newest commits just to have a place for them. I'll do a clean PR tomoor.w

ayende commented:
Closing in favor of a cleanup PR
